include/ginkgo/core/base/scoped_device_id.hpp

```diff
@@ -0,0 +1,228 @@
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_BASE_SCOPED_DEVICE_ID_HPP_
+#define GKO_PUBLIC_CORE_BASE_SCOPED_DEVICE_ID_HPP_
+
+
+#include <memory>
+
+
+namespace gko {
+
+
+class OmpExecutor;
+class CudaExecutor;
+class HipExecutor;
+class DpcppExecutor;
+
+
+namespace detail {
+
+
+/**
+ * A RAII, move-only base class for the scoped device id used for different
+ * executors.
+ */
+class generic_scoped_device_id {
+public:
+    generic_scoped_device_id() = default;
+
+    // TODO: this should be a purely virtual funtion, but somehow that leads to
+    // linker errors
+    // This is explicitly not noexcept, since setting the device id may throw in
+    // the derived classes for CUDA and HIP.
+    // Also, this can't be defaulted, because the default would use
+    // noexcept(true), which leads to the default implementation to be deleted.
+    virtual ~generic_scoped_device_id() noexcept(false){};
+
+    // Prohibit copy construction
+    generic_scoped_device_id(const generic_scoped_device_id& other) = delete;
+
+    // Prohibit copy assignment
+    generic_scoped_device_id& operator=(const generic_scoped_device_id& other) =
+        delete;
+};
+
+
+/**
+ * An implementation of generic_scoped_device_id that does nothing.
+ *
+ * This is used for OmpExecutor and DpcppExecutor, since they don't require
+ * setting a device id.
+ */
+class noop_scoped_device_id : public generic_scoped_device_id {};
+
+
+/**
+ * A scoped device id for CUDA.
+ */
+class cuda_scoped_device_id : public generic_scoped_device_id {
+public:
+    explicit cuda_scoped_device_id(int device_id);
+
+    ~cuda_scoped_device_id() noexcept(false) override;
+
+    cuda_scoped_device_id(cuda_scoped_device_id&& other) noexcept;
+
+    cuda_scoped_device_id& operator=(cuda_scoped_device_id&& other) noexcept;
+
+private:
+    int original_device_id_;
+    bool need_reset_;
+};
+
+
+/**
+ * A scoped device id for HIP.
+ */
+class hip_scoped_device_id : public generic_scoped_device_id {
+public:
+    explicit hip_scoped_device_id(int device_id);
+
+    ~hip_scoped_device_id() noexcept(false) override;
+
+    hip_scoped_device_id(hip_scoped_device_id&& other) noexcept;
+
+    hip_scoped_device_id& operator=(hip_scoped_device_id&& other) noexcept;
+
+private:
+    int original_device_id_;
+    bool need_reset_;
+};
+
+
+}  // namespace detail
```

Comment:
these could be moved to the .cpp file

Reply:
the class definition has to be in some header, because the implementation happens both in core and in cuda/hip. So we could keep it in the public headers or in a new private header. I would lean towards the private header

Reply:
agree :)